src/components/TextInput/BaseTextInput.js

```diff
@@ -147,7 +158,9 @@ class BaseTextInput extends Component {
         if (this.props.onInputChange) {
             this.props.onInputChange(value);
         }
-        this.setState({value});
+        if (!this.props.autoGrow || !this.props.shouldWaitWidthCalculation) {
+            this.setState({value});
+        }
```

Comment:
I have been playing around with this PR to try to avoid [this bug](https://github.com/Expensify/App/pull/18579#issuecomment-1546023693), and I'm pretty sure that this is the change causing it.

When the input is full and you type a new digit, `value` comes with that new digit, and is longer than the permitted value, but we used to set it in the state anyway. Then, `componentDidUpdate` is updated and the value coming from the props is the truncated version. Since the truncated value is different than the `state.value`, we set the truncated value in the state.

With this change, we don't set the longer than permitted value in the state immediately as we used to, and we end up rendering the input with an old `state.value` because we are waiting for the width to be calculated.

I don't understand fully the problem, but I think it is related to this.

Reply:
@aldo-expensify yes you are right that we set state immediately without waiting width calculation - this is the issue we currently have - and to avoid this we should not use state variable for input until width will be calculated

Reply:
@aldo-expensify And also to be clear about render logic:
Why bug with jumping cursor appears - we do not use state value to save longer value - we pass it directly to `src/pages/iou/steps/MoneyRequestAmountPage.js` - inside that page - there is a set state - which cut un needed symbol - put in state and fire re-render.
In BaseTextInput - we receive that value - and compare with value which we have in state
Both are equal - and we do not make any changes in setState - because we get here
```
        if ((_.isUndefined(inputValue) || this.state.value === inputValue) && _.isEqual(prevProps.selection, this.props.selection)) {
            return;
        }
```
But we still in re-render process - because MoneyRequestAmountPage page force re-render by doing setState.
TextInput - get this re-render and check - that value is still the same - but it was detecting keyboard press - and setSelectionChange will be fired here:
```
                        onSelectionChange={(e) => {
                            if (!this.state.shouldUpdateSelection) {
                                return;
                            }
                            this.setState({selection: e.nativeEvent.selection});
                        }}
```
And e.nativeEvent.selection will be last position of the element in input